<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Preprocessing · Juscan.jl</title><meta name="title" content="Preprocessing · Juscan.jl"/><meta property="og:title" content="Preprocessing · Juscan.jl"/><meta property="twitter:title" content="Preprocessing · Juscan.jl"/><meta name="description" content="Documentation for Juscan.jl."/><meta property="og:description" content="Documentation for Juscan.jl."/><meta property="twitter:description" content="Documentation for Juscan.jl."/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">Juscan.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Tutorial</span><ul><li><a class="tocitem" href="../../tutorial/quickstart/">Quick Start</a></li></ul></li><li><span class="tocitem">API</span><ul><li><a class="tocitem" href="../anndata/">Anndata Utils</a></li><li class="is-active"><a class="tocitem" href>Preprocessing</a><ul class="internal"><li><a class="tocitem" href="#quality-control-metrics"><span>quality control metrics</span></a></li><li><a class="tocitem" href="#filter-cells-and-genes"><span>filter cells and genes</span></a></li><li><a class="tocitem" href="#normalization"><span>normalization</span></a></li></ul></li><li><a class="tocitem" href="../tools/">Tools</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">API</a></li><li class="is-active"><a href>Preprocessing</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Preprocessing</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/zehua0417/Juscan.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/zehua0417/Juscan.jl/blob/main/docs/src/api/preprocessing.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Preprocessing-Module"><a class="docs-heading-anchor" href="#Preprocessing-Module">Preprocessing Module</a><a id="Preprocessing-Module-1"></a><a class="docs-heading-anchor-permalink" href="#Preprocessing-Module" title="Permalink"></a></h1><p>Preprocessing is a crucial step in single-cell and multi-omics data analysis. This module provides functions for quality control, filtering, and normalization to ensure that datasets are clean and ready for downstream analyses.</p><p>Quality control (QC) metrics help identify low-quality cells and genes, such as those with extremely high or low counts, high dropout rates, or disproportionate expression levels. By computing these metrics, users can apply appropriate thresholds to retain only reliable data.</p><p>Filtering enables users to remove unwanted cells or genes based on QC criteria, ensuring that low-quality features do not affect downstream analyses.</p><p>Normalization methods adjust for differences in sequencing depth and technical variation, allowing meaningful comparisons across cells and conditions.</p><p>These preprocessing functions ensure that data is well-structured and suitable for clustering, dimensionality reduction, and other analytical tasks.</p><ul><li><a href="#Juscan.Pp.calculate_qc_metrics"><code>Juscan.Pp.calculate_qc_metrics</code></a></li><li><a href="#Juscan.Pp.calculate_qc_metrics!"><code>Juscan.Pp.calculate_qc_metrics!</code></a></li><li><a href="#Juscan.Pp.describe_obs"><code>Juscan.Pp.describe_obs</code></a></li><li><a href="#Juscan.Pp.describe_obs!"><code>Juscan.Pp.describe_obs!</code></a></li><li><a href="#Juscan.Pp.describe_var"><code>Juscan.Pp.describe_var</code></a></li><li><a href="#Juscan.Pp.describe_var!"><code>Juscan.Pp.describe_var!</code></a></li><li><a href="#Juscan.Pp.filter_cells"><code>Juscan.Pp.filter_cells</code></a></li><li><a href="#Juscan.Pp.filter_cells!"><code>Juscan.Pp.filter_cells!</code></a></li><li><a href="#Juscan.Pp.filter_genes"><code>Juscan.Pp.filter_genes</code></a></li><li><a href="#Juscan.Pp.filter_genes!"><code>Juscan.Pp.filter_genes!</code></a></li><li><a href="#Juscan.Pp.normalize_total"><code>Juscan.Pp.normalize_total</code></a></li><li><a href="#Juscan.Pp.normalize_total!"><code>Juscan.Pp.normalize_total!</code></a></li></ul><h2 id="quality-control-metrics"><a class="docs-heading-anchor" href="#quality-control-metrics">quality control metrics</a><a id="quality-control-metrics-1"></a><a class="docs-heading-anchor-permalink" href="#quality-control-metrics" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Juscan.Pp.calculate_qc_metrics" href="#Juscan.Pp.calculate_qc_metrics"><code>Juscan.Pp.calculate_qc_metrics</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">calculate_qc_metrics!(adata; expr_type=&quot;counts&quot;, var_type=&quot;genes&quot;, qc_vars=String[], 
                      percent_top=[50, 100, 200, 500], layer=nothing, use_raw=false, 
                      use_log1p=true, parallel=nothing) -&gt; Tuple{DataFrame, DataFrame}</code></pre><p>Calculate quality control metrics and return.</p><p><strong>Arguments</strong></p><ul><li><code>adata</code>: <code>Muon.AnnData</code>   Annotated data matrix containing single-cell or multi-omics data.</li><li><code>expr_type</code>: <code>String</code> (default: <code>&quot;counts&quot;</code>)   The type of expression data to use, such as raw counts or normalized values.</li><li><code>var_type</code>: <code>String</code> (default: <code>&quot;genes&quot;</code>)   Specifies the variable type, e.g., <code>&quot;genes&quot;</code> for gene expression.</li><li><code>qc_vars</code>: <code>Union{Vector{String}, String}</code> (default: <code>String[]</code>)   A list of quality control variables to consider. If a single string is provided, it will be converted to a vector.</li><li><code>percent_top</code>: <code>Union{Vector{Int}, Nothing}</code> (default: <code>[50, 100, 200, 500]</code>)   A list of top feature (e.g., gene) counts to compute the percentage of total counts for each cell.</li><li><code>layer</code>: <code>Union{String, Nothing}</code> (default: <code>nothing</code>)   Specifies the layer in <code>adata</code> to use. If <code>nothing</code>, the default matrix is used.</li><li><code>use_raw</code>: <code>Bool</code> (default: <code>false</code>)   Whether to use the raw counts matrix (<code>adata.raw</code>) instead of the processed data.</li><li><code>use_log1p</code>: <code>Bool</code> (default: <code>true</code>)   Whether to apply <code>log1p</code> transformation to the data before computing quality control metrics.</li><li><code>parallel</code>: <code>Union{Bool, Nothing}</code> (default: <code>nothing</code>)   <strong>(Deprecated)</strong> This argument is ignored but retained for backward compatibility.</li></ul><p><strong>Returns</strong></p><ul><li>A tuple of two <code>DataFrames.DataFrame</code>:  <ol><li>The first <code>DataFrame</code> contains per-cell quality control metrics.  </li><li>The second <code>DataFrame</code> contains per-gene quality control metrics.</li></ol></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/zehua0417/Juscan.jl/blob/ce5e3e50b2dc0155397cfa109e0837aabf36d090/src/preprocessing/qc.jl#L8-L39">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Juscan.Pp.calculate_qc_metrics!" href="#Juscan.Pp.calculate_qc_metrics!"><code>Juscan.Pp.calculate_qc_metrics!</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">calculate_qc_metrics!(adata; expr_type=&quot;counts&quot;, var_type=&quot;genes&quot;, qc_vars=String[], 
                      percent_top=[50, 100, 200, 500], layer=nothing, use_raw=false, 
                      use_log1p=true, parallel=nothing) -&gt; Nothing</code></pre><p>Calculate and store quality control (QC) metrics directly in the <code>obs</code> and <code>var</code> attributes of <code>adata</code>.</p><p><strong>Arguments</strong></p><ul><li><code>adata</code>: <code>Muon.AnnData</code>   Annotated data matrix containing single-cell or multi-omics data.   The QC metrics will be stored in <code>adata.obs</code> (per-cell) and <code>adata.var</code> (per-gene).</li><li><code>expr_type</code>: <code>String</code> (default: <code>&quot;counts&quot;</code>)   The type of expression data to use, such as raw counts or normalized values.</li><li><code>var_type</code>: <code>String</code> (default: <code>&quot;genes&quot;</code>)   Specifies the variable type, e.g., <code>&quot;genes&quot;</code> for gene expression.</li><li><code>qc_vars</code>: <code>Union{Vector{String}, String}</code> (default: <code>String[]</code>)   A list of quality control variables to consider. If a single string is provided, it will be converted to a vector.</li><li><code>percent_top</code>: <code>Union{Vector{Int}, Nothing}</code> (default: <code>[50, 100, 200, 500]</code>)   A list of top feature (e.g., gene) counts to compute the percentage of total counts for each cell.</li><li><code>layer</code>: <code>Union{String, Nothing}</code> (default: <code>nothing</code>)   Specifies the layer in <code>adata</code> to use. If <code>nothing</code>, the default matrix is used.</li><li><code>use_raw</code>: <code>Bool</code> (default: <code>false</code>)   Whether to use the raw counts matrix (<code>adata.raw</code>) instead of the processed data.</li><li><code>use_log1p</code>: <code>Bool</code> (default: <code>true</code>)   Whether to apply <code>log1p</code> transformation to the data before computing quality control metrics.</li><li><code>parallel</code>: <code>Union{Bool, Nothing}</code> (default: <code>nothing</code>)   <strong>(Deprecated)</strong> This argument is ignored but retained for backward compatibility.</li></ul><p><strong>Returns</strong></p><ul><li><code>Nothing</code>:   This function modifies <code>adata</code> in place by adding QC metrics to <code>adata.obs</code> and <code>adata.var</code>.</li></ul><p><strong>Notes</strong></p><ul><li>Unlike <code>calculate_qc_metrics</code>, which returns QC metrics as <code>DataFrame</code> objects,   this function directly updates <code>adata</code> without returning any values.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/zehua0417/Juscan.jl/blob/ce5e3e50b2dc0155397cfa109e0837aabf36d090/src/preprocessing/qc.jl#L84-L119">source</a></section></article><h3 id="sub-functions"><a class="docs-heading-anchor" href="#sub-functions">sub functions</a><a id="sub-functions-1"></a><a class="docs-heading-anchor-permalink" href="#sub-functions" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Juscan.Pp.describe_obs" href="#Juscan.Pp.describe_obs"><code>Juscan.Pp.describe_obs</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">describe_obs(adata; expr_type=&quot;counts&quot;, var_type=&quot;genes&quot;, qc_vars=String[], 
             percent_top=[50, 100, 200, 500], layer=nothing, use_raw=false, 
             use_log1p=true, X=nothing, parallel=nothing) -&gt; DataFrames.DataFrame</code></pre><p>Compute per-cell quality control (QC) metrics from the expression matrix and return them as a <code>DataFrame</code>.</p><p><strong>Arguments</strong></p><ul><li><code>adata</code>: <code>Muon.AnnData</code>   Annotated data matrix containing single-cell or multi-omics data.</li><li><code>expr_type</code>: <code>String</code> (default: <code>&quot;counts&quot;</code>)   The type of expression data to use, such as raw counts or normalized values.</li><li><code>var_type</code>: <code>String</code> (default: <code>&quot;genes&quot;</code>)   Specifies the variable type, e.g., <code>&quot;genes&quot;</code> for gene expression.</li><li><code>qc_vars</code>: <code>Union{Vector{String}, String}</code> (default: <code>String[]</code>)   A list of quality control variables to consider. If a single string is provided, it will be converted to a vector.</li><li><code>percent_top</code>: <code>Union{Vector{Int}, Nothing}</code> (default: <code>[50, 100, 200, 500]</code>)   A list of top feature (e.g., gene) counts to compute the percentage of total counts for each cell.</li><li><code>layer</code>: <code>Union{String, Nothing}</code> (default: <code>nothing</code>)   Specifies the layer in <code>adata</code> to use. If <code>nothing</code>, the default matrix is used.</li><li><code>use_raw</code>: <code>Bool</code> (default: <code>false</code>)   Whether to use the raw counts matrix (<code>adata.raw</code>) instead of the processed data.</li><li><code>use_log1p</code>: <code>Bool</code> (default: <code>true</code>)   Whether to apply <code>log1p</code> transformation to the data before computing quality control metrics.</li><li><code>X</code>: <code>Union{Nothing, SparseMatrixCSC}</code> (default: <code>nothing</code>)   The expression matrix to use. If <code>nothing</code>, the matrix is determined by <code>_choose_mtx_rep</code>.</li><li><code>parallel</code>: <code>Union{Bool, Nothing}</code> (default: <code>nothing</code>)   <strong>(Deprecated)</strong> This argument is ignored but retained for backward compatibility.</li></ul><p><strong>Returns</strong></p><ul><li><code>DataFrames.DataFrame</code>:   A <code>DataFrame</code> containing per-cell QC metrics, including:<ul><li><code>&quot;n_$(var_type)_by_$(expr_type)&quot;</code>: Number of detected features per cell.</li><li><code>&quot;log1p_n_$(var_type)_by_$(expr_type)&quot;</code>: Log-transformed number of features per cell.</li><li><code>&quot;total_$(expr_type)&quot;</code>: Total expression count per cell.</li><li><code>&quot;log1p_total_$(expr_type)&quot;</code>: Log-transformed total expression count per cell.</li><li><code>&quot;pct_$(expr_type)_in_top_N_$(var_type)&quot;</code>: Percentage of expression from the top <code>N</code> features.</li><li><code>&quot;total_$(expr_type)_$(qc_var)&quot;</code>: Total expression for each QC variable per cell.</li><li><code>&quot;log1p_total_$(expr_type)_$(qc_var)&quot;</code>: Log-transformed total expression for each QC variable per cell.</li><li><code>&quot;pct_$(expr_type)_$(qc_var)&quot;</code>: Percentage of total expression contributed by each QC variable.</li></ul></li></ul><p><strong>Notes</strong></p><ul><li>This function computes per-cell QC metrics based on expression data.   For per-gene metrics, see <code>describe_var</code>.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/zehua0417/Juscan.jl/blob/ce5e3e50b2dc0155397cfa109e0837aabf36d090/src/preprocessing/qc.jl#L193-L237">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Juscan.Pp.describe_obs!" href="#Juscan.Pp.describe_obs!"><code>Juscan.Pp.describe_obs!</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">describe_obs!(adata; expr_type=&quot;counts&quot;, var_type=&quot;genes&quot;, qc_vars=String[], 
              percent_top=[50, 100, 200, 500], layer=nothing, use_raw=false, 
              use_log1p=true, X=nothing, parallel=nothing) -&gt; Nothing</code></pre><p>Compute per-cell quality control (QC) metrics and store them in <code>adata.obs</code>.</p><p><strong>Arguments</strong></p><ul><li><code>adata</code>: <code>Muon.AnnData</code>   Annotated data matrix containing single-cell or multi-omics data.</li><li><code>expr_type</code>: <code>String</code> (default: <code>&quot;counts&quot;</code>)   The type of expression data to use, such as raw counts or normalized values.</li><li><code>var_type</code>: <code>String</code> (default: <code>&quot;genes&quot;</code>)   Specifies the variable type, e.g., <code>&quot;genes&quot;</code> for gene expression.</li><li><code>qc_vars</code>: <code>Union{Vector{String}, String}</code> (default: <code>String[]</code>)   A list of quality control variables to consider. If a single string is provided, it will be converted to a vector.</li><li><code>percent_top</code>: <code>Union{Vector{Int}, Nothing}</code> (default: <code>[50, 100, 200, 500]</code>)   A list of top feature (e.g., gene) counts to compute the percentage of total counts for each cell.</li><li><code>layer</code>: <code>Union{String, Nothing}</code> (default: <code>nothing</code>)   Specifies the layer in <code>adata</code> to use. If <code>nothing</code>, the default matrix is used.</li><li><code>use_raw</code>: <code>Bool</code> (default: <code>false</code>)   Whether to use the raw counts matrix (<code>adata.raw</code>) instead of the processed data.</li><li><code>use_log1p</code>: <code>Bool</code> (default: <code>true</code>)   Whether to apply <code>log1p</code> transformation to the data before computing quality control metrics.</li><li><code>X</code>: <code>Union{Nothing, SparseMatrixCSC}</code> (default: <code>nothing</code>)   The expression matrix to use. If <code>nothing</code>, the matrix is determined by <code>_choose_mtx_rep</code>.</li><li><code>parallel</code>: <code>Union{Bool, Nothing}</code> (default: <code>nothing</code>)   <strong>(Deprecated)</strong> This argument is ignored but retained for backward compatibility.</li></ul><p><strong>Returns</strong></p><ul><li><code>Nothing</code>:   This function modifies <code>adata</code> in place by adding per-cell QC metrics to <code>adata.obs</code>.</li></ul><p><strong>Notes</strong></p><ul><li>This function is similar to <a href="#Juscan.Pp.describe_obs"><code>describe_obs</code></a> but modifies <code>adata</code> directly.   It computes QC metrics using <code>describe_obs</code> and stores the results in <code>adata.obs</code>.  <ul><li>For per-gene QC metrics, see <a href="#Juscan.Pp.describe_var!"><code>describe_var!</code></a>.</li></ul></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/zehua0417/Juscan.jl/blob/ce5e3e50b2dc0155397cfa109e0837aabf36d090/src/preprocessing/qc.jl#L303-L340">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Juscan.Pp.describe_var" href="#Juscan.Pp.describe_var"><code>Juscan.Pp.describe_var</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">describe_var(adata; expr_type=&quot;counts&quot;, var_type=&quot;genes&quot;, layer=nothing, 
             use_raw=false, use_log1p=true, X=nothing) -&gt; DataFrame</code></pre><p>Compute per-gene quality control (QC) metrics and return as a DataFrame.</p><p><strong>Arguments</strong></p><ul><li><code>adata</code>: <code>Muon.AnnData</code>   Annotated data matrix containing single-cell or multi-omics data.</li><li><code>expr_type</code>: <code>String</code> (default: <code>&quot;counts&quot;</code>)   The type of expression data to use, such as raw counts or normalized values.</li><li><code>var_type</code>: <code>String</code> (default: <code>&quot;genes&quot;</code>)   Specifies the variable type, e.g., <code>&quot;genes&quot;</code> for gene expression.</li><li><code>layer</code>: <code>Union{String, Nothing}</code> (default: <code>nothing</code>)   Specifies the layer in <code>adata</code> to use. If <code>nothing</code>, the default matrix is used.</li><li><code>use_raw</code>: <code>Bool</code> (default: <code>false</code>)   Whether to use the raw counts matrix (<code>adata.raw</code>) instead of the processed data.</li><li><code>use_log1p</code>: <code>Bool</code> (default: <code>true</code>)   Whether to apply <code>log1p</code> transformation to the data before computing quality control metrics.</li><li><code>X</code>: <code>Union{Nothing, SparseMatrixCSC}</code> (default: <code>nothing</code>)   The expression matrix to use. If <code>nothing</code>, the matrix is determined by <code>_choose_mtx_rep</code>.</li></ul><p><strong>Returns</strong></p><ul><li><code>DataFrames.DataFrame</code>:   A DataFrame where each row represents a gene (or variable), containing the following QC metrics:<ul><li><code>&quot;n_cells_by_&lt;expr_type&gt;&quot;</code>: Number of cells in which each gene is detected.</li><li><code>&quot;mean_&lt;expr_type&gt;&quot;</code>: Mean expression level of each gene.</li><li><code>&quot;log1p_mean_&lt;expr_type&gt;&quot;</code>: Log-transformed mean expression level.</li><li><code>&quot;pct_dropout_by_&lt;expr_type&gt;&quot;</code>: Percentage of cells in which each gene is not detected.</li><li><code>&quot;total_&lt;expr_type&gt;&quot;</code>: Total expression level of each gene.</li><li><code>&quot;log1p_total_&lt;expr_type&gt;&quot;</code>: Log-transformed total expression level.</li></ul></li></ul><p><strong>Notes</strong></p><ul><li>This function calculates quality control metrics per gene (or other variables).  <ul><li>To store these metrics directly in <code>adata.var</code>, use <a href="#Juscan.Pp.describe_var!"><code>describe_var!</code></a>.  </li><li>To compute per-cell QC metrics, see <a href="#Juscan.Pp.describe_obs"><code>describe_obs</code></a>.</li></ul></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/zehua0417/Juscan.jl/blob/ce5e3e50b2dc0155397cfa109e0837aabf36d090/src/preprocessing/qc.jl#L371-L407">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Juscan.Pp.describe_var!" href="#Juscan.Pp.describe_var!"><code>Juscan.Pp.describe_var!</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">describe_var!(adata; expr_type=&quot;counts&quot;, var_type=&quot;genes&quot;, layer=nothing, 
              use_raw=false, use_log1p=true, X=nothing) -&gt; Nothing</code></pre><p>Compute per-gene quality control (QC) metrics and store them in <code>adata.var</code>.</p><p><strong>Arguments</strong></p><ul><li><code>adata</code>: <code>Muon.AnnData</code>   Annotated data matrix containing single-cell or multi-omics data.</li><li><code>expr_type</code>: <code>String</code> (default: <code>&quot;counts&quot;</code>)   The type of expression data to use, such as raw counts or normalized values.</li><li><code>var_type</code>: <code>String</code> (default: <code>&quot;genes&quot;</code>)   Specifies the variable type, e.g., <code>&quot;genes&quot;</code> for gene expression.</li><li><code>layer</code>: <code>Union{String, Nothing}</code> (default: <code>nothing</code>)   Specifies the layer in <code>adata</code> to use. If <code>nothing</code>, the default matrix is used.</li><li><code>use_raw</code>: <code>Bool</code> (default: <code>false</code>)   Whether to use the raw counts matrix (<code>adata.raw</code>) instead of the processed data.</li><li><code>use_log1p</code>: <code>Bool</code> (default: <code>true</code>)   Whether to apply <code>log1p</code> transformation to the data before computing quality control metrics.</li><li><code>X</code>: <code>Union{Nothing, SparseMatrixCSC}</code> (default: <code>nothing</code>)   The expression matrix to use. If <code>nothing</code>, the matrix is determined by <code>_choose_mtx_rep</code>.</li></ul><p><strong>Returns</strong></p><ul><li><code>Nothing</code>:   This function modifies <code>adata</code> in place by adding per-gene QC metrics to <code>adata.var</code>.</li></ul><p><strong>Notes</strong></p><ul><li>This function is similar to <a href="#Juscan.Pp.describe_var"><code>describe_var</code></a> but modifies <code>adata</code> directly.   It computes QC metrics using <code>describe_var</code> and stores the results in <code>adata.var</code>.  <ul><li>For per-cell QC metrics, see <a href="#Juscan.Pp.describe_obs!"><code>describe_obs!</code></a>.</li></ul></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/zehua0417/Juscan.jl/blob/ce5e3e50b2dc0155397cfa109e0837aabf36d090/src/preprocessing/qc.jl#L452-L482">source</a></section></article><h2 id="filter-cells-and-genes"><a class="docs-heading-anchor" href="#filter-cells-and-genes">filter cells and genes</a><a id="filter-cells-and-genes-1"></a><a class="docs-heading-anchor-permalink" href="#filter-cells-and-genes" title="Permalink"></a></h2><h3 id="filter-cells"><a class="docs-heading-anchor" href="#filter-cells">filter cells</a><a id="filter-cells-1"></a><a class="docs-heading-anchor-permalink" href="#filter-cells" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Juscan.Pp.filter_cells" href="#Juscan.Pp.filter_cells"><code>Juscan.Pp.filter_cells</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">filter_cells(
  data::Muon.AnnData;
  min_counts=nothing, min_genes=nothing,
  max_counts=nothing, max_genes=nothing,
  copy=false) -&gt; Union{Tuple{BitVector, Vector}, Muon.AnnData}</code></pre><p>Filters cells based on the given threshold criteria and returns either a subset mask and count vector or a new filtered <code>Muon.AnnData</code> object.</p><p><strong>Arguments</strong></p><ul><li><code>data::Muon.AnnData</code>: The input single-cell data.</li><li><code>min_counts::Union{Int, Nothing}</code>: Minimum total counts per cell.</li><li><code>min_genes::Union{Int, Nothing}</code>: Minimum number of genes expressed per cell.</li><li><code>max_counts::Union{Int, Nothing}</code>: Maximum total counts per cell.</li><li><code>max_genes::Union{Int, Nothing}</code>: Maximum number of genes expressed per cell.</li><li><code>copy::Bool</code>: If <code>true</code>, returns a filtered copy; otherwise, returns a mask and count vector.</li></ul><p><strong>Returns</strong></p><ul><li>If <code>copy == false</code>, returns a tuple <code>(cell_subset::BitVector, number_per_cell::Vector)</code>.</li><li>If <code>copy == true</code>, returns a filtered <code>Muon.AnnData</code> object.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/zehua0417/Juscan.jl/blob/ce5e3e50b2dc0155397cfa109e0837aabf36d090/src/preprocessing/filter.jl#L13-L33">source</a></section><section><div><pre><code class="language-julia hljs">filter_cells(
  data::AbstractMatrix;
  min_counts=nothing, min_genes=nothing,
  max_counts=nothing, max_genes=nothing
) -&gt; Tuple{BitVector, Vector}</code></pre><p>Filters cells from a count matrix based on the given threshold criteria.</p><p><strong>Arguments</strong></p><ul><li><code>data::AbstractMatrix</code>: Gene expression count matrix with cells as rows.</li><li><code>min_counts::Union{Int, Nothing}</code>: Minimum total counts per cell.</li><li><code>min_genes::Union{Int, Nothing}</code>: Minimum number of genes expressed per cell.</li><li><code>max_counts::Union{Int, Nothing}</code>: Maximum total counts per cell.</li><li><code>max_genes::Union{Int, Nothing}</code>: Maximum number of genes expressed per cell.</li></ul><p><strong>Returns</strong></p><ul><li><code>cell_subset::BitVector</code>: A mask indicating cells that pass the filter.</li><li><code>number_per_cell::Vector</code>: A vector of counts or expressed gene numbers per cell.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/zehua0417/Juscan.jl/blob/ce5e3e50b2dc0155397cfa109e0837aabf36d090/src/preprocessing/filter.jl#L70-L89">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Juscan.Pp.filter_cells!" href="#Juscan.Pp.filter_cells!"><code>Juscan.Pp.filter_cells!</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">filter_cells!(
  data::Muon.AnnData;
  min_counts=nothing, min_genes=nothing,
  max_counts=nothing, max_genes=nothing
) -&gt; Nothing</code></pre><p>Filters cells in-place in a <code>Muon.AnnData</code> object.</p><p><strong>Arguments</strong></p><ul><li><code>data::Muon.AnnData</code>: The input single-cell data.</li><li><code>min_counts::Union{Int, Nothing}</code>: Minimum total counts per cell.</li><li><code>min_genes::Union{Int, Nothing}</code>: Minimum number of genes expressed per cell.</li><li><code>max_counts::Union{Int, Nothing}</code>: Maximum total counts per cell.</li><li><code>max_genes::Union{Int, Nothing}</code>: Maximum number of genes expressed per cell.</li></ul><p><strong>Returns</strong></p><ul><li><code>Nothing</code>. The filtering is applied in-place.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/zehua0417/Juscan.jl/blob/ce5e3e50b2dc0155397cfa109e0837aabf36d090/src/preprocessing/filter.jl#L148-L166">source</a></section></article><h3 id="filter-genes"><a class="docs-heading-anchor" href="#filter-genes">filter genes</a><a id="filter-genes-1"></a><a class="docs-heading-anchor-permalink" href="#filter-genes" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Juscan.Pp.filter_genes" href="#Juscan.Pp.filter_genes"><code>Juscan.Pp.filter_genes</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">filter_genes(
  data::Muon.AnnData;
  min_counts=nothing, min_cells=nothing,
  max_counts=nothing, max_cells=nothing,
  copy=false
) -&gt; Union{Tuple{BitVector, Vector}, Muon.AnnData}</code></pre><p>Filters genes based on the given threshold criteria and returns either a subset mask and count vector or a new filtered <code>Muon.AnnData</code> object.</p><p><strong>Arguments</strong></p><ul><li><code>data::Muon.AnnData</code>: The input single-cell data.</li><li><code>min_counts::Union{Int, Nothing}</code>: Minimum total counts per gene.</li><li><code>min_cells::Union{Int, Nothing}</code>: Minimum number of cells expressing the gene.</li><li><code>max_counts::Union{Int, Nothing}</code>: Maximum total counts per gene.</li><li><code>max_cells::Union{Int, Nothing}</code>: Maximum number of cells expressing the gene.</li><li><code>copy::Bool</code>: If <code>true</code>, returns a filtered copy; otherwise, returns a mask and count vector.</li></ul><p><strong>Returns</strong></p><ul><li>If <code>copy == false</code>, returns a tuple <code>(gene_subset::BitVector, number_per_gene::Vector)</code>.</li><li>If <code>copy == true</code>, returns a filtered <code>Muon.AnnData</code> object.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/zehua0417/Juscan.jl/blob/ce5e3e50b2dc0155397cfa109e0837aabf36d090/src/preprocessing/filter.jl#L201-L222">source</a></section><section><div><pre><code class="language-julia hljs">filter_genes(
  data::AbstractMatrix;
  min_counts=nothing, min_cells=nothing,
  max_counts=nothing, max_cells=nothing
) -&gt; Tuple{BitVector, Vector}</code></pre><p>Filters genes from a count matrix based on the given threshold criteria.</p><p><strong>Arguments</strong></p><ul><li><code>data::AbstractMatrix</code>: Gene expression count matrix with genes as columns.</li><li><code>min_counts::Union{Int, Nothing}</code>: Minimum total counts per gene.</li><li><code>min_cells::Union{Int, Nothing}</code>: Minimum number of cells expressing the gene.</li><li><code>max_counts::Union{Int, Nothing}</code>: Maximum total counts per gene.</li><li><code>max_cells::Union{Int, Nothing}</code>: Maximum number of cells expressing the gene.</li></ul><p><strong>Returns</strong></p><ul><li><code>gene_subset::BitVector</code>: A mask indicating genes that pass the filter.</li><li><code>number_per_gene::Vector</code>: A vector of counts or expressed cell numbers per gene.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/zehua0417/Juscan.jl/blob/ce5e3e50b2dc0155397cfa109e0837aabf36d090/src/preprocessing/filter.jl#L259-L278">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Juscan.Pp.filter_genes!" href="#Juscan.Pp.filter_genes!"><code>Juscan.Pp.filter_genes!</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">filter_genes!(
  data::Muon.AnnData;
  min_counts=nothing, min_cells=nothing,
  max_counts=nothing, max_cells=nothing
) -&gt; Nothing</code></pre><p>Filters genes in-place in a <code>Muon.AnnData</code> object.</p><p><strong>Arguments</strong></p><ul><li><code>data::Muon.AnnData</code>: The input single-cell data.</li><li><code>min_counts::Union{Int, Nothing}</code>: Minimum total counts per gene.</li><li><code>min_cells::Union{Int, Nothing}</code>: Minimum number of cells expressing the gene.</li><li><code>max_counts::Union{Int, Nothing}</code>: Maximum total counts per gene.</li><li><code>max_cells::Union{Int, Nothing}</code>: Maximum number of cells expressing the gene.</li></ul><p><strong>Returns</strong></p><ul><li><code>Nothing</code>. The filtering is applied in-place.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/zehua0417/Juscan.jl/blob/ce5e3e50b2dc0155397cfa109e0837aabf36d090/src/preprocessing/filter.jl#L337-L355">source</a></section></article><h2 id="normalization"><a class="docs-heading-anchor" href="#normalization">normalization</a><a id="normalization-1"></a><a class="docs-heading-anchor-permalink" href="#normalization" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Juscan.Pp.normalize_total" href="#Juscan.Pp.normalize_total"><code>Juscan.Pp.normalize_total</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">normalize_total(
    adata::AnnData;
    target_sum::Union{Real, Nothing}=nothing,
    exclude_highly_expressed::Bool=false,
    max_fraction::Float64=0.05,
    key_added::Union{String, Nothing}=nothing,
    layer::Union{String, Nothing}=nothing,
    layers::Union{String, Vector{String}, Nothing}=nothing,
    layer_norm::Union{String, Nothing}=nothing,
    copy::Bool=false
) -&gt; Union{AnnData, Dict{String, Any}}</code></pre><p>Normalize total counts per cell to a target sum.</p><p><strong>Arguments</strong></p><ul><li><code>adata::AnnData</code>: The single-cell dataset to be normalized.</li><li><code>target_sum::Union{Real, Nothing}=nothing</code>: Target total count per cell. If <code>nothing</code>, normalization is performed to the median of counts per cell.</li><li><code>exclude_highly_expressed::Bool=false</code>: If <code>true</code>, highly expressed genes are excluded from the normalization factor calculation.</li><li><code>max_fraction::Float64=0.05</code>: A gene is considered highly expressed if it accounts for more than <code>max_fraction</code> of the total counts in any cell.</li><li><code>key_added::Union{String, Nothing}=nothing</code>: If provided, the computed normalization factors are stored in <code>adata.obs[key_added]</code>.</li><li><code>layer::Union{String, Nothing}=nothing</code>: The layer to normalize. If <code>nothing</code>, the main count matrix (<code>adata.X</code>) is used.</li><li><code>layers::Union{String, Vector{String}, Nothing}=nothing</code>: Specifies multiple layers to normalize. If <code>&quot;all&quot;</code>, all layers in <code>adata.layers</code> are normalized.</li><li><code>layer_norm::Union{String, Nothing}=nothing</code>: Defines the normalization reference for additional layers. Can be <code>&quot;after&quot;</code>, <code>&quot;X&quot;</code>, or <code>nothing</code>.</li><li><code>copy::Bool=false</code>: If <code>true</code>, returns a copy of <code>adata</code> with normalized counts; otherwise, modifies <code>adata</code> in place.</li></ul><p><strong>Returns</strong></p><ul><li>If <code>copy=true</code>, returns a new <code>AnnData</code> object with normalized counts.</li><li>If <code>copy=false</code>, modifies <code>adata</code> in place and returns a dictionary containing:<ul><li><code>&quot;X&quot;</code>: The normalized count matrix.</li><li><code>&quot;norm_factor&quot;</code>: The computed normalization factors.</li><li>Additional layers if <code>layers</code> is specified.</li></ul></li></ul><p><strong>Notes</strong></p><ul><li>If <code>exclude_highly_expressed=true</code>, genes that exceed <code>max_fraction</code> in any cell are ignored when computing normalization factors.</li><li>Cells with zero counts will generate a warning.</li><li>The function supports layer-wise normalization if <code>layers</code> is specified.</li></ul><p><strong>Example</strong></p><pre><code class="language-julia hljs">adata = AnnData(X)
normalize_total(adata; target_sum=10000)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/zehua0417/Juscan.jl/blob/ce5e3e50b2dc0155397cfa109e0837aabf36d090/src/preprocessing/normalization.jl#L30-L73">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Juscan.Pp.normalize_total!" href="#Juscan.Pp.normalize_total!"><code>Juscan.Pp.normalize_total!</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">normalize_total!(
    adata::AnnData;
    target_sum::Union{Real, Nothing}=nothing,
    exclude_highly_expressed::Bool=false,
    max_fraction::Float64=0.05,
    key_added::Union{String, Nothing}=nothing,
    layer::Union{String, Nothing}=nothing,
    layers::Union{String, Vector{String}, Nothing}=nothing
) -&gt; Nothing</code></pre><p>Normalize total counts per cell <strong>in-place</strong>.</p><p><strong>Arguments</strong></p><ul><li><code>adata::AnnData</code>: The single-cell dataset to be normalized.</li><li><code>target_sum::Union{Real, Nothing}=nothing</code>: Target total count per cell. If <code>nothing</code>, normalization is performed to the median of counts per cell.</li><li><code>exclude_highly_expressed::Bool=false</code>: If <code>true</code>, highly expressed genes are excluded from the normalization factor calculation.</li><li><code>max_fraction::Float64=0.05</code>: A gene is considered highly expressed if it accounts for more than <code>max_fraction</code> of the total counts in any cell.</li><li><code>key_added::Union{String, Nothing}=nothing</code>: If provided, the computed normalization factors are stored in <code>adata.obs[key_added]</code>.</li><li><code>layer::Union{String, Nothing}=nothing</code>: The layer to normalize. If <code>nothing</code>, the main count matrix (<code>adata.X</code>) is used.</li><li><code>layers::Union{String, Vector{String}, Nothing}=nothing</code>: Specifies multiple layers to normalize. If <code>&quot;all&quot;</code>, all layers in <code>adata.layers</code> are normalized.</li></ul><p><strong>Returns</strong></p><ul><li>This function <strong>modifies</strong> <code>adata</code> in place and does not return a new object.</li></ul><p><strong>Notes</strong></p><ul><li>If <code>exclude_highly_expressed=true</code>, genes that exceed <code>max_fraction</code> in any cell are ignored when computing normalization factors.</li><li>Cells with zero counts will generate a warning.</li><li>The function supports layer-wise normalization if <code>layers</code> is specified.</li></ul><p><strong>Example</strong></p><pre><code class="language-julia hljs">adata = AnnData(X)
normalize_total!(adata; target_sum=10000)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/zehua0417/Juscan.jl/blob/ce5e3e50b2dc0155397cfa109e0837aabf36d090/src/preprocessing/normalization.jl#L172-L207">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../anndata/">« Anndata Utils</a><a class="docs-footer-nextpage" href="../tools/">Tools »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.1 on <span class="colophon-date" title="Wednesday 5 March 2025 11:22">Wednesday 5 March 2025</span>. Using Julia version 1.11.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
